### ISSUE TITLE
Pass custom `device_id` to library initialization

### ISSUE THREAD

**Issue by andirsun** [mixpanel/mixpanel-js#422]
Context about why we need this change:

- In my case I am using Next Js + statsig platform to get values for certain A/B tests in our app.
- When the app starts first Next js Middleware is fired (server side), so at this point I have to provide a custom device_id for the statsig to return me the A/B experiment values. At this point I can not initialize mixpanel here because the library is client side and not server side, so I have to create a custom device_id manually to pass to statsig. 
The problem is that I need this device_id to be attached to mixpanel cookie but not with an mixpanel.identify() but as device_id, I think the best solution will be to you allow us to pass device_id in the project initialization so we can control the format of the device_id based on what we need.

**Comment by doc-han** [comment]
`device_id` is unique to the user's device and not the user's account in your DB. I don't understand how you're generating `device_id` server-side(Next.js middleware).

**Comment by andirsun** [comment]
@doc-han hey thanks for the answer!
Basically I am generating a custom uuid in the middleware to act as a device id for statsig.

**Comment by doc-han** [comment]
`device_id` in mixpanel isn't something you're supposed to be worried about changing. It's an anonymous id which isn't tied to anything so let mixpanel handle that. 

A uuid generated by you is supposed to used alongside the `identify` function. Hence, when you generate the uuid in the middleware. share it with the front-end and after mixpanel `init` in the front-end just `identify` with the uuid.

Identify literarily tells mixpanel that the `device_id` and the uuid is the same person. Hence any actions tracked under `device_id` will be associated with the uuid.

**Comment by andirsun** [comment]
Also would be great that I coud pass device_id for my mixpanel implementation on figma plugin. 
https://github.com/mixpanel/mixpanel-js/issues/388

**Comment by Fy-** [comment]
Same here. In my setup, device_id is a cookie set by my Go webserver for all users, including guests, and it is distinct from my user UUID. My Go client-server handles server-side rendering (SSR) for a website. It would be beneficial to allow passing a custom  function so that the same device ID can be used consistently on both client and server sides (e.g., `get_device_id: () => getCookie('__did')`).

This is ugly but it works (after initial load or if you do the page event manually):

``` javascript
// with cookie_name: '__mp_data' in mixpanel.init opts.

const mpData = cookies.mp___mp_data
if (mpData) {
  const mpDataJson = JSON.parse(decodeURIComponent(mpData))
  mpDataJson.$device_id = cookies.__did
  mixpanel.register({
    $device_id: cookies.__did,
  })
  mpCookie.set('mp___mp_data', JSON.stringify(mpDataJson), {
    expires: new Date(new Date().getTime() + 1000 * 60 * 60 * 24 * 365),
    domain: '.domain.com',
  })
}
```
Note: The `__did` cookie is used solely for device identification (not user identification) by my webserver and is also passed along for server-side events. For instance, a user may have multiple devices, each with different `__did` values, but the UUID will always remain the same for identifying the user.

**Comment by fulstadev** [comment]
Came across this, and I need the exact same thing, basically to be able to track page views on the server-side. According to [this](https://docs.mixpanel.com/docs/tracking-best-practices/server-side-best-practices#tracking-page-views) and [this](https://docs.mixpanel.com/docs/tracking-methods/id-management/identifying-users-simplified#server-side-identity-management), this means that the anonymous ID / device ID must be generated on the server-side, especially for users which arrive on your website for the first time, in which case the client-side SDK was not processed before the page load hence no device ID is available.

For this scenario, it is also essential to be able to pass the anonymous ID / device ID manually to the mixpanel JS SDK client. I think that the biggest problem with this is that the server-side mixpanel libraries [do not provide the possibility](https://docs.mixpanel.com/docs/tracking-best-practices/server-side-best-practices#identifying-users) to generate anonymous IDs out of the box. There are actually very basic vendors out there offer a functionality in their server-side SDK to set the anonymous ID on the server-side, via a simple function call, requiring the anonymous ID as single parameter. If you call it, what it does under the hood is:

- it sets the anonymous ID to the one you provided. If your provided value is `null`, the server-side SDK automatically generates a random UUID, which will then be used as the anonymous ID.
- Any server-side SDK tracking request are related to the anonymous ID resulting from the previous step.
- it (once per running server-side request) also adds a `Set-Cookie` header to the servers' response headers, through which it will automatically forward the anonymous ID back to the client when the server responds. Let's say the name of that cookie is `X`.
- The client-side Js SDK is then constructed to automatically pick up the presence of a cookie named `X`, and in that case automatically assigns the anonymous ID used by the Js SDK client to it.

There are also SDK implementations of other vendors that directly use the same cookie. So instead of generating a `Set-Cookie` header for a random cookie name `X` as above, in the case of mixpanel, the server-side example above would generate a `Set-Cookie` header for the `mp_<project_token>_mixpanel` cookie to automatically override the client-side cookie accordingly.

A few questions here: 

1. @doc-han You mention to simply call `identify()` with the device ID / anonymous ID obtained from the server, on the client. This would correspond to assigning both the device ID and also the distinct ID to future requests, until the identity is reset. Is that correct? In this case, would identity merge / resolution still work properly? E.g. are the following two events merged into the same profile :

- event with device ID `1234-acbd-5678-efghijk` and distinct ID `1234-acbd-5678-efghijk`
- event with device ID `1234-acbd-5678-efghijk` and distinct ID `456`

?

2. @doc-han I was a little by your suggestion of calling `identify()` with an anoymous ID, as [mixpanels' docs](https://docs.mixpanel.com/docs/tracking-methods/sdks/javascript#identify) say:

> We recommend against calling .identify() for anonymous visitors to your site.

Could you thus maybe elaborate on your suggestion?

**Comment by emezeske** [comment]
Hey, +1 on this. Would really love a supported way to set the device_id at init (or a set_device_id method).

Our situation: we generate our own first-party anonymous id in a cookie and use it as the $device_id on all of our server-side event ingestion. We recently added browser Session Replay, and we want those replays to line up with the same identity as our server-side events, including for anonymous visitors who haven't logged in yet. We're on Simplified ID Merge and our server already stamps $device_id = our cookie value, so really the browser SDK just needs to use that same value as its $device_id.

The existing options don't quite get us there:

- identify() sets $user_id, not $device_id, so passing our anonymous id there files it as a logged-in user and forks identity (and it's meant for the real user id anyway).
- register({ '$device_id': ourId }) sort of works, since $device_id is a registered superprop, but it's undocumented and feels fragile. We're not confident Session Replay's identity association actually honors it, and it seems like the kind of thing that could silently break on a future SDK version.
- Linking through a shared $user_id at login only helps once someone logs in. Anonymous replays still land on a separate profile from our server-side anonymous events and never merge.

So a first-class device_id init option (or a setter) would let us keep one consistent identity across server, desktop, and browser replay without any of those workarounds.

Is there something technically challenging about making this feature available?
